### PR TITLE
Debug traces + CB

### PR DESCRIPTION
**Pull request by CedricGuillemet**
Documentation will be part of BabylonNative.

2 levels of logging
# Code logging
Works are a console log. Enable Debug Trace (off by default) and set an outputter(none set by default) then call Trace.
```
SetTraceOutput([](const char *trace){
printf("%s\n", trace);
});
SetDebugTrace(true);
...
Trace("Error detected.");
```

# Babylon code instrumentation
Babylon code is instrumented and will use Trace function described earlier IF `JSRUNTIMEHOST_DEBUG_TRACE` preprocessor is defined. If it's not (default), no log will happen and no performance impact.

To summarize, user can log in his own code with Babylon logging API. Babylon Native and JSRuntimeHost code is instrumented but will only log if  `JSRUNTIMEHOST_DEBUG_TRACE` preprocessor is defined and logging is enabled and an outputer function is set.

Thread safety in the logging output is responsability of the user.

### CODE REVIEW

File: Core/Foundation/Include/Babylon/DebugTrace.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+#include <functional>
+
+#define DEBUG_TRACE_NOOP(...)
```
Comment (bghgary) [inline Core/Foundation/Include/Babylon/DebugTrace.h:4]: Why do you need this?

File: Core/Foundation/Include/Babylon/DebugTrace.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+#define _STRINGIFY(x) #x
+#define _TOSTRING(x) _STRINGIFY(x)
+#define _DEBUG_TRACE(_format, ...) Babylon::DebugTrace::Trace(__FILE__ "(" _TOSTRING(__LINE__) "): " _format "\n", ##__VA_ARGS__);
```
Comment (bghgary) [inline Core/Foundation/Include/Babylon/DebugTrace.h:8]: nit: Why does `_format` have an underscore?

File: Core/Foundation/Include/Babylon/DebugTrace.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#endif // JSRUNTIMEHOST_DEBUG_TRACE
+
+namespace Babylon
+{
+    namespace DebugTrace
+    {
+        // off by default.
+        void SetDebugTrace(bool enabled);
+        // equivalent of a printf that will use TraceOutput function to log. Need DebugTrace to be enabled.
+        void Trace(const char* _format, ...);
```
Comment (bghgary) [inline Core/Foundation/Include/Babylon/DebugTrace.h:23]: Ditto on `_format`

File: Core/Foundation/Include/Babylon/DebugTrace.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#else
+#   define DEBUG_TRACE(...) DEBUG_TRACE_NOOP()
+#endif // JSRUNTIMEHOST_DEBUG_TRACE
+
+namespace Babylon
+{
+    namespace DebugTrace
+    {
+        // off by default.
+        void SetDebugTrace(bool enabled);
```
Comment (bghgary) [inline Core/Foundation/Include/Babylon/DebugTrace.h:21]: Maybe a better name?
```suggestion
        void EnableDebugTrace(bool enabled);
```

File: Core/Foundation/Include/Babylon/DebugTrace.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        // equivalent of a printf that will use TraceOutput function to log. Need DebugTrace to be enabled.
+        void Trace(const char* _format, ...);
+        // set the function used to output.
+        void SetTraceOutput(std::function<void(const char* trace)> traceOutputFunction);
```
Comment (bghgary) [inline Core/Foundation/Include/Babylon/DebugTrace.h:25]: Do we need two functions? It seems like we can use one function for enabling/setting the callback. If someone wants to enable, they have to provide a callback?

File: Core/Foundation/Source/DebugTrace.cpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        static bool s_debugTraceEnabled{};
+        static std::function<void(const char* output)> s_traceOutputFunction;
+        static void TraceArgs(const char* format, va_list args)
+        {
+            char temp[8192];
+            auto len = vsnprintf(temp, sizeof(temp), format, args);
+            temp[len] = '\0';
+            s_traceOutputFunction(temp);
+        }
```
Comment (bghgary) [inline Core/Foundation/Source/DebugTrace.cpp:18]: Anonymous namespace doesn't need to be under the Babylon namespace.

Also, `static` is redundant inside an anonymous namespace.

File: Core/Foundation/Include/Babylon/DebugTrace.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+#define _STRINGIFY(x) #x
+#define _TOSTRING(x) _STRINGIFY(x)
+#define _DEBUG_TRACE(_format, ...) Babylon::DebugTrace::Trace(__FILE__ "(" _TOSTRING(__LINE__) "): " _format "\n", ##__VA_ARGS__); 
+
+#if JSRUNTIMEHOST_DEBUG_TRACE
+#   define DEBUG_TRACE _DEBUG_TRACE
+#else
+#   define DEBUG_TRACE(...) DEBUG_TRACE_NOOP()
+#endif // JSRUNTIMEHOST_DEBUG_TRACE
```
Comment (bghgary) [inline Core/Foundation/Include/Babylon/DebugTrace.h:14]: I don't think this is enough. We probably need something that allows us to check if debug trace is enabled so that people can wrap more complex debug code in an `#ifdef`. I guess we can check `JSRUNTIMEHOST_DEBUG_TRACE`, but maybe this name isn't good?

File: Core/Foundation/Source/DebugTrace.cpp [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+{
+    namespace
+    {
+        static bool s_debugTraceEnabled{};
+        static std::function<void(const char* output)> s_traceOutputFunction;
+        static void TraceArgs(const char* format, va_list args)
+        {
+            char temp[8192];
+            auto len = vsnprintf(temp, sizeof(temp), format, args);
+            temp[len] = '\0';
```
Comment (bghgary) [inline Core/Foundation/Source/DebugTrace.cpp:16]: `vsnprintf` may return negative values if an error occurs. Maybe should check that.

Comment (CedricGuillemet) [reply Core/Foundation/Source/DebugTrace.cpp]: Fixed.

File: Core/Foundation/Include/Babylon/DebugTrace.h (hunk 7 above)
Comment (CedricGuillemet) [reply Core/Foundation/Include/Babylon/DebugTrace.h]: I've added DEBUG_RUN that takes a lambda. It works. Isn't it over kill compared to something like `assert` ? Do you have a use case for that? `JSRUNTIMEHOST_DEBUG_TRACE` might indeed not be a good name :) Naming is hard!

Comment (CedricGuillemet) [reply Core/Foundation/Include/Babylon/DebugTrace.h]: using BABYLON_DEBUG_TRACE  instead. macro with lambda removed.

File: Core/Foundation/Source/DebugTrace.cpp [hunk 9]
```diff
@@ -0,0 +1,52 @@
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <functional>
+
+namespace
+{
+    bool s_debugTraceEnabled{};
```
Comment (bghgary) [inline Core/Foundation/Source/DebugTrace.cpp:8]: nit: `s_` is for static members of a class. This should probably be `g_` for global variables.

File: Tests/UnitTests/Android/app/src/main/cpp/JNI.cpp [hunk 10]
```diff
@@ -21,6 +21,9 @@ Java_com_jsruntimehost_unittests_Native_javaScriptTests(JNIEnv* env, jclass claz
 
     android::global::Initialize(javaVM, context);
 
+    Babylon::DebugTrace::EnableDebugTrace(true);
+    Babylon::DebugTrace::SetTraceOutput([](const char* trace) { printf("%s\n", trace); });
```
Comment (bghgary) [inline Tests/UnitTests/Android/app/src/main/cpp/JNI.cpp:25]: Ideally should flush after printf to avoid buffering. Maybe use std::cout instead which has a flush function.

File: Tests/UnitTests/Linux/App.cpp [hunk 11]
```diff
@@ -1,6 +1,11 @@
 #include "../Shared/Shared.h"
+#include "Babylon/DebugTrace.h"
+#include <cstdio>
 
 int main()
 {
+    Babylon::DebugTrace::EnableDebugTrace(true);
+    Babylon::DebugTrace::SetTraceOutput([](const char* trace) { printf("%s\n", trace); });
```
Comment (bghgary) [inline Tests/UnitTests/Linux/App.cpp:8]: Ditto

**Review by bghgary — APPROVED** [review]

File: Core/Foundation/Source/DebugTrace.cpp (hunk 9 above)
Comment (CedricGuillemet) [reply Core/Foundation/Source/DebugTrace.cpp:8]: Fixed.

File: Tests/UnitTests/Android/app/src/main/cpp/JNI.cpp (hunk 10 above)
Comment (CedricGuillemet) [reply Tests/UnitTests/Android/app/src/main/cpp/JNI.cpp:25]: Fixed.

File: Tests/UnitTests/Linux/App.cpp (hunk 11 above)
Comment (CedricGuillemet) [reply Tests/UnitTests/Linux/App.cpp:8]: Fixed.
